core/server/common/src/main/java/alluxio/master/journal/raft/transport/CopycatGrpcClient.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.journal.raft.transport;
+
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.CopycatMessageServerGrpc;
+import alluxio.grpc.GrpcChannel;
+import alluxio.grpc.GrpcChannelBuilder;
+import alluxio.grpc.GrpcServerAddress;
+import alluxio.security.user.UserState;
+
+import io.atomix.catalyst.concurrent.ThreadContext;
+import io.atomix.catalyst.transport.Address;
+import io.atomix.catalyst.transport.Client;
+import io.atomix.catalyst.transport.Connection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Copycat transport {@link Client} implementation that uses Alluxio gRPC.
+ */
+public class CopycatGrpcClient implements Client {
+  private static final Logger LOG = LoggerFactory.getLogger(CopycatGrpcClient.class);
+
+  /** Alluxio configuration. */
+  private AlluxioConfiguration mConf;
+  /** Authentication user. */
+  private UserState mUserState;
+
+  /** Created channels. */
+  private Map<Address, GrpcChannel> mChannels;
```

Comment:
Do we expect this client to connect to make multiple connections to the same addresses or make multiple connections to different addresses? It seems more intuitive for each client to be to one address and share channels between clients.

Reply:
Copycat try to utilize single client for making connections to multiple servers. So pooling here is in line with their usage and as a result we will be authenticating each high-level Copycat client.

Also, actual managed channel pooling is still in effect when using GrpcChannelBuilder. So there is only 1 physical connection is opened to same target host.